tests/transformers/tests/generation/test_utils.py

```diff
@@ -1491,503 +1241,253 @@ def test_assisted_decoding_sample(self):
-                            sequences=input_ids,
-                            scores=scores,
-                            attentions=decoder_attentions,
-                            hidden_states=decoder_hidden_states,
-                            past_key_values=model_kwargs.get("past_key_values"),
-                        )
-                else:
-                    return input_ids
+            self._check_outputs(output_assisted, input_ids, config, use_cache=True)
+
+    @pytest.mark.generate
+    def test_prompt_lookup_decoding_stops_at_eos(self):
+        # This test ensures that the prompt lookup generation stops at eos token and does not suggest more tokens
+        # (see https://github.com/huggingface/transformers/pull/31301)
 
-            model.assisted_decoding = MethodType(assisted_decoding, model)
+        # The main idea is to have an ngram (unigram in our case) that is repeated twice in the input ids.
+        # First time at the very end, so input ends with the unigrams, and second any arbitrary location.
+        # Also, we need an EOS token which will be injected just after the arbitrary located ngram.
+        # We verify that PLD will not copy and propose candidated that contain an EOS token, even if there are overlapping ngrams
+        # in input ids. Otherwise a proposed EOS along with the trailing (ngrams-1) tokens might be accepted by the target model.
+        # That seems as if the model "generated" and EOS but didn't stop from user's perspective
 
-            #######################################################################
+        input_ids = torch.randint(1, 50, (1, 10), device=torch_device)  # generate inputs in range from 1-50
+        arbitrary_ngram = 51  # this is the arbitrary ngram, specifically chosen OOV to prevent flaky tests
+        input_ids[:, 3] = arbitrary_ngram  # set pre-eos to arbitrary_ngram which is for sure not present in inputs
+        input_ids[:, -1] = arbitrary_ngram  # put arbitrary_ngram in the end for the necessary match to happen
 
-            output_assisted = model.generate(input_ids, attention_mask=attention_mask, **generation_kwargs)
+        eos_token_id = torch.tensor([0], device=torch_device)
+        input_ids[:, 4] = eos_token_id  # inject eos-token-id in input ids so that it is located after arbitrary_ngram
 
-            self._check_outputs(output_assisted, input_ids, model.config, use_cache=True)
+        # init cand geenerator with max_matching_ngram_size=1 to match per-token
+        candidate_generator = PromptLookupCandidateGenerator(
+            eos_token_id=eos_token_id, num_output_tokens=4, max_matching_ngram_size=1
+        )
+        output_prompt_lookup = candidate_generator.get_candidates(input_ids)[0]
+
+        # PLD shouldn't propose any new tokens based on eos-match
+        self.assertTrue(output_prompt_lookup.shape[-1] == 10)
 
+    @pytest.mark.generate
     def test_generate_with_head_masking(self):
```

Comment:
@malkomes there is a minor correction in this test, not on your modification, but please correct line 1516
`output_attentions=True,` -> `output_attentions=self.has_attentions,`

Reply:
Although I might be wrong. I have to check again. It seems we do not use `self.has_attentions` at every places

Reply:
There is also `**inputs_dict` missing in `out = model.generate(` call

Reply:
great catch!